CMakeLists.txt

```diff
@@ -123,22 +128,33 @@ if(NOT OPTIMIZE STREQUAL "off")
     endif()
 
     if(OPTIMIZE STREQUAL "portable" OR OPTIMIZE STREQUAL "fastbuild")
-      message(STATUS "Enabling SS2 CPU optimizations (>= Pentium 4)")
-      # SSE and SSE2 are core instructions on x64
-      # and consequently raise a warning message from compiler with this flag on x64.
       if(NOT CMAKE_SIZEOF_VOID_P EQUAL 8)
+        # Target architecture is x86 with SSE and SSE2
+        message(STATUS "x86 Enabling SS2 CPU optimizations (>= Pentium 4)")
+        # Define gcc/clang style defines for SSE and SSE2 for compatibility
+        add_compile_definitions("__SSE__" "__SSE2__")
+        # Set compiler option for SSE/SSE2
         add_compile_options(/arch:SSE2)
       endif()
-      add_compile_definitions("__SSE__" "__SSE2__")
     elseif(OPTIMIZE STREQUAL "native")
       message("Enabling native optimizations for ${CMAKE_SYSTEM_PROCESSOR}")
```

Comment:
This message is wrong, because msvc has no concept of native flags. We should warn about that and fall back to portable.

Reply:
native is intended to enable AVX and such like /arch:[AVX|AVX2|AVX512]

If you have interest, you can integrate this: 
https://github.com/intel/media-lab/blob/master/video_analytics_Intel_GPU/cmake/CPUID.cmake

It is used here with the intel compiler flags, This can be adopted for Microsoft. 
https://github.com/intel/media-lab/blob/master/video_analytics_Intel_GPU/cmake/OptimizationFlags.cmake

Reply:
Maybe this is an alternative: https://github.com/miurahr/cmake-optimize-architecture-flag
It's CMake only (without dependency to the CPUID executable) and supports ARM (Linux/Android only) too, but not 32 Bit x86.

Reply:
We're not making 32 bit x86 Windows builds anymore anyway.

Reply:
This uses a hard coded flag list for known architectures. 
https://github.com/miurahr/cmake-optimize-architecture-flag/blob/d83310a05ce14bb7cdb08bcf1e44c78e3d6707f9/Modules/GetCPUSIMDFeatures/GetX64SIMDFeatures.cmake
This list is already outdated. I guess this is not a good fit for us. 

The intel solution is here much better, because it fits to all CPUs and Microsoft takes care that this is really the case. 
https://docs.microsoft.com/de-de/cpp/intrinsics/cpuid-cpuidex?view=msvc-160

@JoergAtGithub Do you have interest to integrate that? 
If not, we should fail the build, if one is using "native" and msvc.

Reply:
It implements both:
-Detecting the host processor architecture for native optimization 
-Allows to specify a target processor architecture by name

The Intel solution is also outdated. I noticed, that MSVC supports the AVX512 instruction set since version 2017, but there's just a comment, that MSVC doesn't support it.

An alternative could be also: https://github.com/VcDevel/Vc/tree/master
It seems to be wider adopted, but it's missing ARM NEON support.

Reply:
> @JoergAtGithub Do you have interest to integrate that?

I'm considering this, but not in this PR. Here I want just fix the bug, that the MSCV compiler uses SSE/SSE2 optimization and the code assumes the precision of non SSE/SSE2 Floating-point number, because the code relies on preprocessor defines which are compiler specific.

Reply:
Ok, thank you. 
So this can be merged or do you plan to improve the messages for the intermediate state?

Reply:
> https://github.com/VcDevel/Vc/tree/master

needs also be updated for new CPUs.

This is a different type of outdated.
The Intel solution asks the cpu flags directly. If Intel or Amd releases a new CPU it will still report whether it has avx registers or not.
Enabling a new flag supported for the compiler is on top. In the hardcoded solution no flag is set and it falls back to portable.

Do you know how "native" is implemented on mingw gcc?

Reply:
It used also __cpuid

Reply:
I added a dummy code line where the user can set the flags for his system. Followed by an error message, that the user has to do this.